hidi/matrix.py

```diff
@@ -226,6 +227,82 @@ def transform(self, M,  **kwargs):
         return self.keras_model, kwargs
 
 
+class KerasKfoldTransform(Transform):
+    """
+    Generalized transform for Keras algorithm with k fold cross validation
+    evaluation
+    """
+    def __init__(self, keras_model, validation_matrix, tensorboard_log_dir,
+                 tts_seed=42, tt_split=0.25, kfold_n_splits=10,
+                 kfold_seed=42, kfold_shuffle=True,
+                 classification=False, **keras_kwargs):
+        self.keras_model = keras_model
+        self.keras_kwargs = keras_kwargs
+        self.validation_matrix = validation_matrix
+
+        self.kfold_n_splits = kfold_n_splits
+        self.kfold_seed = kfold_seed
+        self.kfold_shuffle = kfold_shuffle
+
+        self.tts_seed = tts_seed
+        self.tt_split = tt_split
+
+        self.log_dir = tensorboard_log_dir
+
+        self.classification = classification
+
+        if 'item_id' in validation_matrix.columns:
+            self.validation_matrix.set_index('item_id', inplace=True)
+
+    def transform(self, M,  **kwargs):
+        """
+        Takes a Takes a dataframe that has :code:`item_id` index, other
+        'features' columns for prediction, and applies a Keras sequential
+        model to it.
+
+        The model will be evaluated using tensorboard and returns None
+
+        """
+        rows, columns = M.shape
+        embedding = M.merge(self.validation_matrix, left_index=True,
+                            right_index=True)
+        embedding = embedding.values
+
+        x_train, x_test, y_train, y_test = train_test_split(
+            embedding[:, :columns], embedding[:, columns:],
+            random_state=self.tts_seed, test_size=self.tt_split)
+
+        if self.classification:
+            kfold = StratifiedKFold(n_splits=self.kfold_n_splits,
+                                    random_state=self.kfold_seed,
+                                    shuffle=self.kfold_shuffle)
+        else:
+            kfold = KFold(n_splits=self.kfold_n_splits,
+                          random_state=self.kfold_seed,
+                          shuffle=self.kfold_shuffle)
+
+        n_fold = 1
+        for train_index, test_index in kfold.split(x_train, y_train):
+            self.keras_model.fit(
+                x_train[train_index], y_train[train_index],
+                validation_data=[x_test, y_test],
+                **self.keras_kwargs)
+
+            files = os.listdir(self.log_dir)
+            newfile_path = os.path.join(self.log_dir,
+                                        '{0}_fold'.format(n_fold))
+            if not os.path.exists(newfile_path):
+                os.makedirs(newfile_path)
+            for file in files:
+                if file.startswith('event'):
+                    move(os.path.join(self.log_dir, file),
+                         newfile_path)
+
+            n_fold += 1
+
+        return None, kwargs
```

Comment:
Does it make sense for this to return the trained model, similar to our non-kfold keras transform?

https://github.com/VEVO/hidi/blob/e266f945f0868fffeeb157da0abf232a16444b9d/hidi/matrix.py#L226